### PR TITLE
Pretransformations to use destination passing style, before upstream bufferization.

### PR DESCRIPTION
**Pull request by MaheshRavishankar**
The upstream ComprehensiveBufferization pass prefers a more
destination-passing style of code, i.e. it has a way to know before
hand what the output buffer to use is for in-place
operations. Specifically, the use of `init_tensor` to conjure a tensor
(even though it might be fully over-written before its values are
used) results in an allocation. It is easier to do a
pre-transformation that removes these `init_tensor` operations and
replaces them with a `flow.dispatch.tensor.load` of the output where
the final result goes into.
The current IREE bufferization handles this. This patch reuses the
relevant parts of IREE bufferization to effect this transformation.

### CODE REVIEW

**Comment by MaheshRavishankar** [comment]
Requires https://reviews.llvm.org/D113522

**Review by antiagainst — APPROVED** [review]
Looks it's mostly moving code around; so approve.

**Review by matthias-springer — COMMENTED** [review]
With the recent changes in Comprehensive Bufferize, there is an easier way to do this. The idea is same: Replace the InitTensorOp with a DispatchTensorLoadOp. But some of the code required to do this is now in ComprehensiveBufferize.cpp and can be reused here.

I'm uploading a PR later today. That being said, I'm not doing anything with TensorExpandShapeOp or casts etc., so I'm wondering if this change is handling cases that I did not think of. Are there any additional test cases (apart from iree_comprehensive_bufferize.mlir)?

File: iree/compiler/Codegen/Common/ConvertToDestinationPassingStylePass.cpp [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+      }
+      if (Operation *definingOp = val.getDefiningOp()) {
+        b.clone(*definingOp, indexValMap);
+      }
+      clonedVals.push_back(indexValMap.lookup(val));
+    }
+    return clonedVals;
+  };
+  SmallVector<OpFoldResult> loadOffsets, loadSizes, loadStrides;
+  loadOffsets = cloneIndexValues(op.getMixedOffsets());
```
Comment (matthias-springer) [inline iree/compiler/Codegen/Common/ConvertToDestinationPassingStylePass.cpp:119]: Why are you cloning these? Is it possible to use `getMixedOffsets()` directly?

File: iree/compiler/Codegen/Common/ConvertToDestinationPassingStylePass.cpp [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+static Value getTensorLoadOpForTensorStoreOp(OpBuilder &b, Operation *storeOp) {
+  SmallVector<Value, 4> operandsOfLoadOp;
+  auto op = cast<OffsetSizeAndStrideOpInterface>(storeOp);
+  Value target, source;
+  std::tie(source, target) =
+      TypeSwitch<Operation *, std::tuple<Value, Value>>(op)
+          .Case<IREE::Flow::DispatchTensorStoreOp>([&](auto storeOp) {
+            return std::make_tuple(storeOp.value(), storeOp.target());
+          })
+          .Case<tensor::InsertSliceOp>([&](auto storeOp) {
```
Comment (matthias-springer) [inline iree/compiler/Codegen/Common/ConvertToDestinationPassingStylePass.cpp:69]: Functionality for removing `InsertSliceOp`-anchored `InitTensorOp`s has landed in MLIR ComprehensiveBufferize. We can use that directly now. I'll prepare a commit that enables it and we just need to add a unit test.

Comment (matthias-springer) [reply iree/compiler/Codegen/Common/ConvertToDestinationPassingStylePass.cpp:69]: Such `InitTensorOp`s are replaced with `ExtractSliceOp`s. So it's a bit different from what this is doing. The resulting IR should be the same though.

**Comment by MaheshRavishankar** [comment]
> With the recent changes in Comprehensive Bufferize, there is an easier way to do this. The idea is same: Replace the InitTensorOp with a DispatchTensorLoadOp. But some of the code required to do this is now in ComprehensiveBufferize.cpp and can be reused here.
> 
> I'm uploading a PR later today. That being said, I'm not doing anything with TensorExpandShapeOp or casts etc., so I'm wondering if this change is handling cases that I did not think of. Are there any additional test cases (apart from iree_comprehensive_bufferize.mlir)?

Thats great. This is just another pre-pass that allows a bit more of a glide path (cause I know this works for IREEs cases). You can use the tests here as examples. Once your PR works and we can just bufferize directly then we can just drop this pass. Its not doing any harm.

File: iree/compiler/Codegen/Common/ConvertToDestinationPassingStylePass.cpp (hunk 1 above)
Comment (MaheshRavishankar) [reply iree/compiler/Codegen/Common/ConvertToDestinationPassingStylePass.cpp:119]: Some times these offset values might be defined "after" the point at which I need to insert the `flow.dispatch.load.tensor` (or `tensor.extract_slice` in the upstream case). So cloning just makes it easier. Needs to work for dynamic cases as well.

File: iree/compiler/Codegen/Common/ConvertToDestinationPassingStylePass.cpp (hunk 2 above)
Comment (MaheshRavishankar) [reply iree/compiler/Codegen/Common/ConvertToDestinationPassingStylePass.cpp:69]: Yeah it should end up with the same

**Pull request merged**